### PR TITLE
Add Reply to Message Support

### PR DESCRIPTION
**Pull request by iamstevedavis**
- Updated to Node 14
- Support Text Replied
 - All replies (unless error) respond with `Thank you for your confirmation!`
 - Interpreter is not supported; A new ticket needs to be created for that (it's basically a new feature)

### CODE REVIEW

**Comment by iamstevedavis** [comment]
ToDo: Add Tests

File: src/api/handlers/twilio/postReply.js [hunk 1]
```diff
@@ -1,15 +1,138 @@
...
 
-  const replyText = TemplatesModel.generateReply(language);
+  return convertedReply;
+};
 
-  const reply = getMessageResponse(replyText);
+export default async (req, res) => {
+  let patientPhoneNumber = req?.body?.From;
+  const messageFromPatient = req?.body?.Body;
+  const lowercaseMessageFromPatient = messageFromPatient.toLowerCase();
```
Comment (Wiles) [inline src/api/handlers/twilio/postReply.js:35]: This should be moved to after the null check is done on line 41.

File: src/api/handlers/twilio/postReply.js [hunk 2]
```diff
@@ -1,15 +1,138 @@
...
 
-  res.set("Content-Type", "text/xml");
-  res.send(reply);
+  let appointments;
+  try {
+    appointments = await db("appointments")
+      .select("appointmentId", "language")
+      .where({ patientPhoneNumber, appointmentIsConfirmed: 0 });
+  } catch (error) {
+    // Can't store reply without patientId
```
Comment (Wiles) [inline src/api/handlers/twilio/postReply.js:54]: `patientId`?

File: src/api/handlers/twilio/postReply.js [hunk 3]
```diff
@@ -1,15 +1,138 @@
...
+  try {
+    switch (lowercaseMessageFromPatient) {
+      case "yes":
+        for (const appointment of appointments) {
+          await db("replies")
+            .insert({
+              phoneNumber: patientPhoneNumber,
+              body: messageFromPatient,
+              appointmentId: appointment.appointmentId,
+              time: moment().format("YYYY-MM-DD HH:mm:ss"),
```
Comment (Wiles) [inline src/api/handlers/twilio/postReply.js:87]: Any reason for `moment().format("YYYY-MM-DD HH:mm:ss")` over `new Date()`?

File: src/api/handlers/twilio/postReply.js [hunk 4]
```diff
@@ -1,15 +1,138 @@
...
+    return res.send(ourResponse);
+  }
 
-  res.set("Content-Type", "text/xml");
-  res.send(reply);
+  let appointments;
+  try {
+    appointments = await db("appointments")
+      .select("appointmentId", "language")
+      .where({ patientPhoneNumber, appointmentIsConfirmed: 0 });
```
Comment (Wiles) [inline src/api/handlers/twilio/postReply.js:52]: Why `0`/`1` over `false`/`true`?

File: Dockerfile [hunk 5]
```diff
@@ -1,4 +1,4 @@
-FROM node:12.13.1-alpine as build
+FROM node:14.11.0-alpine as build
```
Comment (mhakmi) [inline Dockerfile:1]: Should update `.nvmrc` file to point to v14?

File: src/api/handlers/twilio/postReply.js [hunk 6]
```diff
@@ -1,15 +1,157 @@
...
+      break;
+    }
+  }
 
-  const replyText = TemplatesModel.generateReply(language);
+  return convertedReply;
+};
+
+export async function getAppointments(patientPhoneNumber) {
+  return await db("appointments")
```
Comment (mhakmi) [inline src/api/handlers/twilio/postReply.js:33]: I think this should return the `DB` promise without `async/await`, as it's already handled below at line `52`

File: src/api/handlers/twilio/postReply.js [hunk 7]
```diff
@@ -1,15 +1,157 @@
...
+  let ourResponse;
+
+  if (!messageFromPatient || messageFromPatient === "") {
+    ourResponse = getMessageResponse(
+      "We are sorry, we could not process your response. Please call Sanctuary Refugee Health Centre (Dr. Michael Stephenson) at 226-336-1321"
+    );
+    return ourResponse;
+  }
+
+  const lowercaseMessageFromPatient = messageFromPatient.toLowerCase();
```
Comment (mhakmi) [inline src/api/handlers/twilio/postReply.js:48]: It'd good to trim whitespaces from the `messageFromPatient`, as it's common that phone keyboards add spaces

File: src/api/handlers/twilio/postReply.js [hunk 8]
```diff
@@ -1,15 +1,157 @@
...
+        }
+        break;
+      case "no":
+        for (const appointment of appointments) {
+          await db("replies")
+            .insert({
+              phoneNumber: patientPhoneNumber,
+              body: messageFromPatient,
+              appointmentId: appointment.appointmentId,
+              time: moment().format("YYYY-MM-DD HH:mm:ss"),
```
Comment (mhakmi) [inline src/api/handlers/twilio/postReply.js:105]: Should this be `new Date()` instead? to coincide with the case above at line `87`

File: src/api/handlers/twilio/postReply.js [hunk 9]
```diff
@@ -1,15 +1,157 @@
...
+          await db("appointments")
+            .where({ appointmentId: appointment.appointmentId })
+            .update({
+              appointmentIsConfirmed: false,
+            })
+            .transacting(trx);
+        }
+        break;
+      // Handle interpreter?
+      default:
```
Comment (mhakmi) [inline src/api/handlers/twilio/postReply.js:117]: How about handling the case when the user reply did not recognize as one of the cases above (yes, no or interpreter) by replying with a message asking the user to reply with a valid response. something like:

> We are sorry, your response not valid. Could you please send (Yes) to confirm the appointment, (No) to decline the appointment, or (Interpreter) to request an interpretation aid with the appointment.

File: src/services/twilioClient.js [hunk 10]
```diff
@@ -1,16 +1,15 @@
...
-  twilioConfig.accountSid,
-  twilioConfig.authToken
-);
-const MessagingResponse = require("twilio").twiml.MessagingResponse;
-const countryCode = "+1";
-
-export const webhook = twilio.webhook;
+export const webhook = () => {
+  const twilio = require("twilio");
+  twilio.webhook;
```
Comment (mhakmi) [inline src/services/twilioClient.js:5]: Should this be returned?

**Review by mhakmi — CHANGES_REQUESTED** [review]
Looks perfect! 👍  Just left a few comments/suggestions.

File: src/api/handlers/twilio/postReply.js (hunk 6 above)
Comment (iamstevedavis) [reply src/api/handlers/twilio/postReply.js:33]: Good idea, this function is still a work in progress. Might not be needed.

File: src/api/handlers/twilio/postReply.js [hunk 11]
```diff
@@ -1,15 +1,157 @@
...
+    );
+    return ourResponse;
+  }
+
+  const lowercaseMessageFromPatient = messageFromPatient.toLowerCase();
+
+  let appointments;
+  try {
+    appointments = await getAppointments(patientPhoneNumber);
+  } catch (error) {
```
Comment (Wiles) [inline src/api/handlers/twilio/postReply.js:53]: You should either log `error` or pass it up. Doing nothing makes it much harder to debug problems.

File: src/api/handlers/twilio/postReply.js (hunk 9 above)
Comment (iamstevedavis) [reply src/api/handlers/twilio/postReply.js:88]: That case is handled above with a message already. We can change it of course. Interpreter needs to be handled still as part of a separate ticket and we have no translations.

Comment (mhakmi) [reply src/api/handlers/twilio/postReply.js:100]: Ah right! sounds good 👍

**Review by mhakmi — CHANGES_REQUESTED** [review]
It looks awesome! I just noticed that the Twilio middleware webhook not invoked properly [at this 36 in ./src/api/index.js](https://github.com/SanctuaryRefugeeHealth/sanctuary-patient-api/blob/master/src/api/index.js#L36), as Twilio returns a factory that returns middleware function, so it changed to be `webhook({ protocol: config.protocol })()`. Just this change and this PR should be ready to ship :)

I tried to push it, but looks like I don't have permission for that.

**Comment by iamstevedavis** [comment]
Like this? `api.post("/twilio/reply", webhook({ protocol: config.protocol })(), postReply);`

**Comment by mhakmi** [comment]
> Like this? `api.post("/twilio/reply", webhook({ protocol: config.protocol })(), postReply);`
Yes, exactly!

**Comment by iamstevedavis** [comment]
> > Like this? `api.post("/twilio/reply", webhook({ protocol: config.protocol })(), postReply);`
> Yes, exactly! 

How did you test it?

**Comment by mhakmi** [comment]
> > > Like this? `api.post("/twilio/reply", webhook({ protocol: config.protocol })(), postReply);`
> > > Yes, exactly!
> 
> How did you test it?

Ah, good question! I used [ngrok](https://ngrok.com/download) to expose my localhost to the public internet, then configured my trial Twilio account with a webhook URL points to my local env. Basically, you'll need to run the api on your env, then run `ngrok` like this in the terminal `./ngrok HTTP 8080`, then it'll provide you with a URL be used as a webhook in Twilio settings. from more detailed steps, I followed [this article](https://www.twilio.com/blog/2013/10/test-your-webhooks-locally-with-ngrok.html#:~:text=In%20your%20Twilio%20account%20dashboard,Save%20your%20changes.)

**Comment by iamstevedavis** [comment]
> > > > Like this? `api.post("/twilio/reply", webhook({ protocol: config.protocol })(), postReply);`
> > > > Yes, exactly!
> > 
> > 
> > How did you test it?
> 
> Ah, good question! I used [ngrok](https://ngrok.com/download) to expose my localhost to the public internet, then configured my trial Twilio account with a webhook URL points to my local env. Basically, you'll need to run the api on your env, then run `ngrok` like this in the terminal `./ngrok HTTP 8080`, then it'll provide you with a URL be used as a webhook in Twilio settings. from more detailed steps, I followed [this article](https://www.twilio.com/blog/2013/10/test-your-webhooks-locally-with-ngrok.html#:~:text=In%20your%20Twilio%20account%20dashboard,Save%20your%20changes.)

alright that is what I have been doing. I am getting an error, the app won't start with your change.

**Comment by iamstevedavis** [comment]
@mhakmi can we move that change into a new issue? I don't really have time to figure out the issue right now. @Wiles said it could be related to an old version of the module.

**Comment by mhakmi** [comment]
> > > > > Like this? `api.post("/twilio/reply", webhook({ protocol: config.protocol })(), postReply);`
> > > > > Yes, exactly!
> > > 
> > > 
> > > How did you test it?
> > 
> > 
> > Ah, good question! I used [ngrok](https://ngrok.com/download) to expose my localhost to the public internet, then configured my trial Twilio account with a webhook URL points to my local env. Basically, you'll need to run the api on your env, then run `ngrok` like this in the terminal `./ngrok HTTP 8080`, then it'll provide you with a URL be used as a webhook in Twilio settings. from more detailed steps, I followed [this article](https://www.twilio.com/blog/2013/10/test-your-webhooks-locally-with-ngrok.html#:~:text=In%20your%20Twilio%20account%20dashboard,Save%20your%20changes.)
> 
> alright that is what I have been doing. I am getting an error, the app won't start with your change.

Weird! could you please post the error here? I checked it last and was working. I suspect there are some changes after pushing the prettier changes, I'll check it now

**Comment by Wiles** [comment]
It has to do with the change to `twilio-node` from `3.39.0` to `3.49.4`. I cannot find any documentation regarding this change but I can confirm that without the extra `()` the endpoint is not properly secure on the new version of the library.

**Comment by mhakmi** [comment]
@iamstevedavis Okay, that's it! My bad, no need to update anything. Apparently, I tested it before you push the [prettier commit](https://github.com/SanctuaryRefugeeHealth/sanctuary-patient-api/pull/63/commits/3ee848d51639a30c4a117cf5724476c70f658df1) which the webhook were exported like this 
`export const webhook = () => { const webhook = require("twilio"); return webhook };`
but after the changes, it became `export const webhook = twilio.webhook;`

**Review by mhakmi — APPROVED** [review]
It's ready to ship 🚢 🎉

**Pull request merged**